File: src/main/java/com/github/wolray/line/io/FieldSelector.java
package com.github.wolray.line.io;

import java.lang.reflect.Field;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.function.Predicate;

/**
 * @author wolray
 */
class FieldSelector {
    boolean pojo;
    String[] use;
    String[] omit;
    String useRegex;
    String omitRegex;

    static FieldSelector of(Fields fields) {
        FieldSelector selector = new FieldSelector();
        if (fields != null) {
            selector.pojo = fields.pojo();
            selector.use = fields.use();
            selector.omit = fields.omit();
            selector.useRegex = fields.useRegex();
            selector.omitRegex = fields.omitRegex();
        }
        return selector;
    }

    static Predicate<Field> toPredicate(Fields fields) {
        if (fields == null) {
            return f -> true;
        }
        return of(fields).toPredicate();
    }

    Predicate<Field> toPredicate() {
        if (use != null && use.length > 0) {
            Set<String> set = new HashSet<>(Arrays.asList(use));
            return f -> set.contains(f.getName());
        }
        if (omit != null && omit.length > 0) {
            Set<String> set = new HashSet<>(Arrays.asList(omit));
            return f -> !set.contains(f.getName());
        }
        if (useRegex != null && !useRegex.isEmpty()) {
            return f -> f.getName().matches(useRegex);
        }
        if (omitRegex != null && !omitRegex.isEmpty()) {
            return f -> !f.getName().matches(omitRegex);
        }
        return f -> true;
    }
}

File: src/main/java/com/github/wolray/line/io/FieldSelector.kt
package com.github.wolray.line.io

import com.github.wolray.line.io.EmptyScope.ifNotEmpty
import java.lang.reflect.Field

/**
 * @author wolray
 */
class FieldSelector {
    var pojo = false
    var use: Array<String>? = null
    var omit: Array<String>? = null
    var useRegex: String? = null
    var omitRegex: String? = null

    fun toTest(): (Field) -> Boolean {
        use.ifNotEmpty {
            val set = toSet()
            return { set.contains(it.name) }
        }
        omit.ifNotEmpty {
            val set = toSet()
            return { !set.contains(it.name) }
        }
        useRegex.ifNotEmpty {
            val regex = toRegex()
            return { it.name.matches(regex) }
        }
        omitRegex.ifNotEmpty {
            val regex = toRegex()
            return { !it.name.matches(regex) }
        }
        return { true }
    }

    companion object {
        @JvmStatic
        fun of(fields: Fields?) = FieldSelector().apply {
            fields?.also {
                pojo = it.pojo
                use = it.use
                omit = it.omit
                useRegex = it.useRegex
                omitRegex = it.omitRegex
            }
        }
    }
}
